src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php

```diff
@@ -79,22 +59,12 @@ public function isPasswordValid($user, string $plainPassword): bool
         return $hasher->verify($user->getPassword(), $plainPassword, $salt);
     }
 
-    /**
-     * @param PasswordAuthenticatedUserInterface $user
-     */
-    public function needsRehash($user): bool
+    public function needsRehash(PasswordAuthenticatedUserInterface $user): bool
     {
         if (null === $user->getPassword()) {
             return false;
         }
 
-        if (!$user instanceof PasswordAuthenticatedUserInterface) {
-            if (!$user instanceof UserInterface) {
-                throw new \TypeError(sprintf('Expected an instance of "%s" as first argument, but got "%s".', UserInterface::class, get_debug_type($user)));
-            }
-            trigger_deprecation('symfony/password-hasher', '5.3', 'The "%s()" method expects a "%s" instance as first argument. Not implementing it in class "%s" is deprecated.', __METHOD__, PasswordAuthenticatedUserInterface::class, get_debug_type($user));
-        }
-
```

Comment:
There is no union type on this method, this goes outside the scope of this PR :)